libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRoutesTracker.kt

```diff
@@ -0,0 +1,104 @@
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+/***
+ * Keeps track of primary and alternatives routes to notify in [NewFasterRouteObserver] that a
+ * new faster route is available.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class FasterRoutesTracker internal constructor(
+    private val mapboxNavigation: MapboxNavigation,
+    private val fasterRouteTrackerCore: FasterRouteTrackerCore,
+    private val scope: CoroutineScope
+) {
+
+    private var previousRouteUpdateProcessing: Job? = null
+    private val newFasterRoutesObservers = CopyOnWriteArraySet<NewFasterRouteObserver>()
+
+    private val internalObserver = RoutesObserver { result: RoutesUpdatedResult ->
+        previousRouteUpdateProcessing?.cancel()
+        previousRouteUpdateProcessing = scope.launch {
+            val fasterRouteTrackerResult = fasterRouteTrackerCore.findFasterRouteInUpdate(
+                result,
+                mapboxNavigation.getAlternativeMetadataFor(result.navigationRoutes)
+            )
+            when (fasterRouteTrackerResult) {
+                is FasterRouteResult.NewFasterRouteFound -> newFasterRouteFound(
+                    NewFasterRoute(
+                        fasterRouteTrackerResult.route,
+                        fasterRouteTrackerResult.fasterThanPrimaryBy,
+                        fasterRouteTrackerResult.alternativeId
+                    )
+                )
+                FasterRouteResult.NoFasterRoute -> {}
+            }
+        }
+    }
+
+    init {
+        mapboxNavigation.registerRoutesObserver(internalObserver)
+    }
+
+    fun registerNewFasterRouteObserver(observer: NewFasterRouteObserver) {
+        newFasterRoutesObservers.add(observer)
+    }
+
+    fun unregisterNewFasterRouteObserver(observer: NewFasterRouteObserver) {
+        newFasterRoutesObservers.remove(observer)
+    }
+
+    /***
+     * Sets faster route as primary to [MapboxNavigation]
+     */
+    @UiThread
+    fun acceptFasterRoute(newFasterRoute: NewFasterRoute) {
+        val currentRoutes = mapboxNavigation.getNavigationRoutes()
+        if (currentRoutes.contains(newFasterRoute.fasterRoute)) {
+            mapboxNavigation.setNavigationRoutes(
+                listOf(newFasterRoute.fasterRoute) +
+                    currentRoutes.filterNot { it == newFasterRoute.fasterRoute }
+            )
+        } else {
+            logE("Ignoring accepted faster route as it's not present in current routes")
+        }
+    }
+
+    /***
+     * Remembers faster route as declined to not offer similar to this one again.
+     */
+    @UiThread
+    fun declineFasterRoute(newFasterRoute: NewFasterRoute) {
+        fasterRouteTrackerCore.fasterRouteDeclined(
+            newFasterRoute.alternativeId,
+            newFasterRoute.fasterRoute
+        )
+    }
+
+    /***
+     * Stops observation and processing routes from [MapboxNavigation]
+     */
+    @UiThread
+    fun destroy() {
```

Comment:
I think we're expecting too much from the user of this API by exposing the whole object, and also making the user maintain the lifecycle. Why don't we always collect the routes in the `FasterRouteTrackerCore` and process them only if at least one observer is registered, and at the same time hide the lifecycle under the `MapboxNavigation` hood?

Reply:
> Why don't we always collect the routes in the FasterRouteTrackerCore and process them only if at least one observer is registered, and at the same time hide the lifecycle under the MapboxNavigation hood?

I wanted to do the same at the beginning, but then I thought "why do I need add such a complex initialisation if I'm not even sure that the feature makes sense?". So I decided to keep things simpler while feature is experimental.

> I think we're expecting too much from the user of this API by exposing the whole object, and also making the user maintain the lifecycle.

After seeing the comments I realised that the lifecycle might be confusing. It's an experimental feature and I can simplify lifecycle by removing `destroy`. Without calling `destroy` the feature lives as long as `MapboxNavigaiton`. I think it's okay for an experimental feature. If we decide that the feature makes sense, I will implement sophisticated system with initialisation and destructions of faster routes.

Reply:
:+1: